src/api/encode.rs

```diff
@@ -0,0 +1,175 @@
+// LNP/BP Rust Library
+// Written in 2020 by
+//     Dr. Maxim Orlovsky <orlovsky@pandoracore.com>
+// Encoding strategies concept developed by
+//     Martin Habovštiak <martin.habovstiak@gmail.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the MIT License
+// along with this software.
+// If not, see <https://opensource.org/licenses/MIT>.
+
+
+use std::vec::IntoIter;
+use std::convert::TryFrom;
+use zmq::Message;
+
+use bitcoin::consensus::encode::{
+    serialize as consensus_serialize,
+    deserialize as consensus_deserialize
+};
+
+use crate::bp::ShortId;
+#[cfg(feature="use-rgb")]
+use crate::csv;
+use crate::csv::{network_serialize, network_deserialize};
+use super::{Multipart, Error};
+
+
+// 1. Encoding messages
+pub trait MessageEncode where Self: Sized {
+    type Error: std::error::Error;
+    fn into_message(self) -> Message;
+    fn try_from_message(message: Message) -> Result<Self, Self::Error>;
+}
+
+// This is a trick for rust compiler helping to distinguish types implementing
+// mutually-exclusive traits (required until negative trait impls will be there)
+// Implemented after concept by Martin Habovštiak <martin.habovstiak@gmail.com>
+mod strategy {
+    use core::marker::PhantomData;
+
+    // Defining strategies:
+    pub enum BitcoinConsensus { }
+    #[cfg(feature="use-rgb")]
+    pub enum RGBStrategy { }
+    pub enum Native { }
+    pub trait Other { type Strategy; }
+
+    #[repr(transparent)]
+    pub struct Holder<T, S>(T, PhantomData<S>);
+    impl<T, S> Holder<T, S> {
+        pub fn new(val: T) -> Self { Self(val, PhantomData::<S>::default()) }
+        pub fn into_inner(self) -> T { self.0 }
+    }
+}
+
+// 1.1. Auto impl for bitcoin-serialized types
+impl<T> MessageEncode for strategy::Holder<T, strategy::BitcoinConsensus>
+    where T: bitcoin::consensus::encode::Encodable +
+             bitcoin::consensus::encode::Decodable {
+    type Error = Error;
+    fn into_message(self) -> Message {
+        Message::from(consensus_serialize(&self.into_inner()))
+    }
+    fn try_from_message(message: Message) -> Result<Self, Self::Error> {
+        Ok(Self::new(consensus_deserialize(&message)?))
+    }
+}
+
+// 1.2. Auto impl for client-validation-serialized types
+#[cfg(feature="use-rgb")]
+impl<T> MessageEncode for strategy::Holder<T, strategy::RGBStrategy>
+    where T: csv::serialize::Network {
+    type Error = Error;
+    fn into_message(self) -> Message {
+        Message::from(network_serialize(&self.into_inner())
+            .expect("Commitment serialize failed"))
```

Comment:
Not strongly related, but serialization failing is quite uncommon thing (unless `serde` is use because of its dynamic nature). Since this is RGB serialization, it'd be nice to try remove the possibility of failure.

Reply:
In RGB, serialization may fail. After Peter Todd suggestion we have very strict bounds to all possible values; and if some of arrays is going outsized (like it must be `u8` and rust supports only `usize`-based limits for arrays) serialization MAY fail, indicating data integrity problem, that can't be handled at any other level. 

Also, `::network_serialize` is a high-level function, using IO-based write procedure underneath, which is certainly have to be held at upstream level. Even `::network_serialize` itself uses `io::Cursor`, which may fail due to a memory allocation problem, for instance.

Reply:
Interesting, looks like a pretty good idea. In that case, it might be better to change the signature to `fn to_massage(&self) -> Result<Message, Self::Error>` or  `fn to_massage(&self) -> Result<Message, Self::SerializationError>` to pass the error to the caller. But panicking here is not entirely bad for Rust standards (panic if the problem was caused by the programmer) either. Also notice that I changed it to take `&self` and the name to start with `to` per Rust conventions.

AFAIK `Cursor` doesn't attempt to handle memory allocations by returning `Err`, it's the sub-optimal definition of `Read` and `Write` that force it into returning `io::Result`, but it never returns `Error` anyway. That's pretty much why I wrote [`genio`](https://crates.io/crates/genio).

Reply:
The difference between `network_serialize` and serialization for ZMQ-based APIs/IPCs is the following: network serialization is not an implementation-specific and is a LNPBP standard important to maintain the network "consensus" (all peers can exchange the data in the same way). On the other hand, ZMQ APIs are "vendor-specific"; they are not a part of P2P network consensus and are related either to IPCs between daemons or client-server connections. 

In serialization the error means that the data integrity or code implementation is broken. For P2P networking, it's important to gracefully handle such errors. In APIs, no reason of returning error to the client; we just need to panic and shut down so the developers know that they have to change the code — or the daemon has to restart to do the data integrity checks.

Reply:
In that case, wouldn't the possibility of an API problem bringing down the whole node be problematic?

Reply:
It must not be problematic by the design of the node: (1) docker container is always restarted on crash, (2) all operations are transactional and (3) node state integrity is maintained with them and persisted with each state update. Also not the whole node will crash, but a single daemon, like `wired` and this will be detected.

I just can't see how the caller can handle an API serialization error other then crashing as well, so there is no need in propagating it.

Reply:
OK, thanks for explaining!